script.js

```diff
@@ -0,0 +1,942 @@
+      }
+      break;
+    }
+    case variables.specialKeys.TAB: {
+      event.preventDefault();
+      const start = text.selectionStart;
+      const end = text.selectionEnd;
+
+      const strBefore = text.value.substring(0, start);
+      const strAfter = text.value.substring(end);
+      const tab = '\t';
+
+      // set textarea value to: text before caret + tab + text after caret
+      text.value = `${strBefore}${tab}${strAfter}`;
+
+      // put caret at right position
+      text.selectionStart = start + tab.length;
+      text.selectionEnd = text.selectionStart;
+      break;
+    }
+    case variables.specialKeys.CAPS_LOCK: {
+      variables.isCapslock = !variables.isCapslock;
+      changeCase();
+      break;
+    }
+    case variables.specialKeys.SHIFT: {
+      if (!event.repeat) {
+        variables.isShift = !variables.isShift;
+        changeCase();
+        changeOnShift();
+      }
+      break;
+    }
+    case variables.specialKeys.ENTER: {
+      event.preventDefault();
+      const start = text.selectionStart;
+      const end = text.selectionEnd;
+
+      const strBefore = text.value.substring(0, start);
+      const strAfter = text.value.substring(end);
+      const newline = '\n';
+
+      // set textarea value to: text before caret + newline + text after caret
+      text.value = `${strBefore}${newline}${strAfter}`;
+
+      // put caret at right position
+      text.selectionStart = start + newline.length;
+      text.selectionEnd = text.selectionStart;
+      break;
+    }
+    case variables.specialKeys.SPACE: {
+      event.preventDefault();
+
+      const start = text.selectionStart;
+      const end = text.selectionEnd;
+
+      const strBefore = text.value.substring(0, start);
+      const strAfter = text.value.substring(end);
+      const symbol = ' ';
+
+      // set textarea value to: text before caret + tab + text after caret
+      text.value = `${strBefore}${symbol}${strAfter}`;
+
+      // put caret at right position
+      text.selectionStart = start + symbol.length;
+      text.selectionEnd = text.selectionStart;
+      break;
+    }
+    case variables.specialKeys.ARROW_LEFT: {
+      event.preventDefault();
+      const start = text.selectionStart;
+      const end = text.selectionEnd;
+
+      // check if selection presents
+      if (start !== end) {
+        // remove selection
+        text.selectionEnd = text.selectionStart;
+      } else if (start !== 0) {
+        // handle out of left boundaries
+        text.selectionStart = start - 1;
+        text.selectionEnd = text.selectionStart;
+      }
+      break;
+    }
+    case variables.specialKeys.ARROW_RIGHT: {
+      event.preventDefault();
+      const start = text.selectionStart;
+      const end = text.selectionEnd;
+
+      // check if selection presents
+      if (start !== end) {
+        // remove selection
+        text.selectionStart = text.selectionEnd;
+      } else if (end !== text.value.length) {
```

Comment:
What happend if property 'text.value' will be undefined? I think you should check this case. 
Better to write condition like this (text.value && text.value.length). And you can move the condition to outside variable for example const isTextEmpty = text.value && text.value.length;